pkg/controller/beat/common/reconcile.go

```diff
@@ -0,0 +1,138 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package common
+
+import (
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	beatv1beta1 "github.com/elastic/cloud-on-k8s/pkg/apis/beat/v1beta1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/daemonset"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/deployment"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/reconciler"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/pointer"
+)
+
+func reconcilePodVehicle(podTemplate corev1.PodTemplateSpec, params DriverParams) *reconciler.Results {
+	results := reconciler.NewResult(params.Context)
+	if err := params.Validate(); err != nil {
+		return results.WithError(err)
+	}
+
+	spec := params.Beat.Spec
+	name := Name(params.Beat.Name, spec.Type)
+
+	var toDelete runtime.Object
+	var reconciliationFunc func(params ReconciliationParams) (int32, int32, error)
+	switch {
+	case spec.DaemonSet != nil:
+		reconciliationFunc = reconcileDaemonSet
+		toDelete = &v1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: params.Beat.Namespace,
+			},
+		}
+	case spec.Deployment != nil:
+		reconciliationFunc = reconcileDeployment
+		toDelete = &v1.DaemonSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: params.Beat.Namespace,
+			},
+		}
+	}
+
+	ready, desired, err := reconciliationFunc(ReconciliationParams{
+		client:      params.Client,
+		beat:        params.Beat,
+		podTemplate: podTemplate,
+	})
+	if err != nil {
+		return results.WithError(err)
+	}
+
+	// clean up the other one
+	if err := params.Client.Delete(toDelete); err != nil && !apierrors.IsNotFound(err) {
```

Comment:
We could get (from the cache) then delete if necessary to avoid an unnecessary call to the apiserver.